File: project_euler_unified/problem_193/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    // Count squarefree numbers below 2^50
    // Q(N) = sum_{k=1}^{sqrt(N)} mu(k) * floor(N/k^2)
    const long long N = (1LL << 50) - 1;
    const int SQ = 1 << 25; // sqrt(2^50) = 2^25

    // Sieve Mobius function
    vector<int> mu(SQ + 1, 1);
    vector<bool> is_prime(SQ + 1, true);
    vector<int> primes;

    // We'll use a sieve that tracks the smallest prime factor
    // and whether a number is divisible by p^2 for any prime p
    // Simple approach: for each prime p, multiply mu by -1 for multiples of p,
    // set mu=0 for multiples of p^2

    for (int i = 2; i <= SQ; i++) {
        if (is_prime[i]) {
            primes.push_back(i);
            // i is prime
            for (long long j = i; j <= SQ; j += i) {
                is_prime[j] = (j == i);
                mu[j] *= -1;
            }
            long long i2 = (long long)i * i;
            for (long long j = i2; j <= SQ; j += i2) {
                mu[j] = 0;
            }
        }
    }

    long long ans = 0;
    for (int k = 1; k <= SQ; k++) {
        if (mu[k] != 0) {
            ans += (long long)mu[k] * (N / ((long long)k * k));
        }
    }

    cout << ans << endl;
    return 0;
}

File: project_euler_unified/problem_193/solution.py
"""
Problem 193: Squarefree Numbers

Count squarefree numbers below 2^50.
Uses: Q(N) = sum_{k=1}^{sqrt(N)} mu(k) * floor(N/k^2)
"""

def solve():
    N = (1 << 50) - 1
    SQ = 1 << 25  # 2^25 = 33554432

    # Sieve Mobius function up to SQ
    mu = [0] * (SQ + 1)
    mu[1] = 1

    # Use a sieve: for each prime p, multiply mu by -1 for multiples,
    # set mu=0 for multiples of p^2
    is_prime = bytearray(b'\x01') * (SQ + 1)
    is_prime[0] = is_prime[1] = 0
    mu = [1] * (SQ + 1)

    for i in range(2, SQ + 1):
        if is_prime[i]:
            for j in range(i, SQ + 1, i):
                if j != i:
                    is_prime[j] = 0
                mu[j] *= -1
            i2 = i * i
            for j in range(i2, SQ + 1, i2):
                mu[j] = 0

    ans = 0
    for k in range(1, SQ + 1):
        if mu[k] != 0:
            ans += mu[k] * (N // (k * k))

    return ans

print(solve())
